### ISSUE TITLE
Use StringMatcher for route path and header matching

### ISSUE THREAD

**Issue by htuch** [envoyproxy/envoy#8392]
As per `#next-major-version` comments in https://github.com/envoyproxy/envoy/blob/master/api/envoy/api/v2/route/route.proto. We should ideally do this in v2 with new fields and then deprecate the existing matchers, allowing for auto-deletion in v3.

**Comment by htuch** [comment]
@mattklein123 do you know who might own this work?

**Comment by mattklein123** [comment]
I will assign this to myself but leave marked help wanted in case someone else wants to step up.

**Comment by yittg** [comment]
@htuch @mattklein123 Looks like that we only need to adapt the structured `StringMatcher` in the configuration parser, maybe I can take it.

**Comment by yittg** [comment]
One more thing, maybe we can do it along with refactoring the matcher in `jwt_authn` if needed.

**Comment by htuch** [comment]
@yittg yes, but there is some subtle issues, e.g. how do you handle the query fragment of the path in a generic way. The ideal is we figure this out and create a parallel match config in v2 (or v3 if this comes after 1.12.0), then deprecate the existing fields.

**Comment by htuch** [comment]
@mattklein123 are you planning on landing this for 1.13.0 and v3, or do we punt?

**Comment by mattklein123** [comment]
No not going to get to this, moving out.
